apps/backend/src/modules/user-routes/user-route.service.ts

```diff
@@ -0,0 +1,190 @@
+import { HTTPCode } from "~/libs/enums/enums.js";
+import { getCurrentIsoDate } from "~/libs/helpers/helpers.js";
+import { type LineStringGeometry, type Service } from "~/libs/types/types.js";
+
+import { type RouteService } from "../routes/route.service.js";
+import {
+	UserRouteExeptionMessage,
+	UserRouteStatus,
+} from "./libs/enums/enum.js";
+import { UserRouteError } from "./libs/exeptions/exeptions.js";
+import {
+	type UserRouteResponseDto,
+	type UserRouteStatusType,
+} from "./libs/types/type.js";
+import { UserRouteEntity } from "./user-route.entity.js";
+import { type UserRouteRepository } from "./user-route.repository.js";
+
+class UserRouteService implements Service {
+	private routeService: RouteService;
+
+	private userRouteRepository: UserRouteRepository;
+
+	public constructor(
+		userRouteRepository: UserRouteRepository,
+		routeService: RouteService,
+	) {
+		this.userRouteRepository = userRouteRepository;
+		this.routeService = routeService;
+	}
+
+	public async create(payload: {
+		routeId: number;
+		userId: number;
+	}): Promise<UserRouteResponseDto> {
+		const { routeId, userId } = payload;
+
+		const { geometry } = await this.routeService.findById(routeId);
+
+		const createdData = UserRouteEntity.initializeNew({
+			actualGeometry: geometry,
+			plannedGeometry: geometry,
+			routeId,
+			status: UserRouteStatus.NOT_STARTED,
+			userId,
+		});
+
+		await this.ensureIsNotDuplicateRoute(routeId, userId);
+
+		const createdRoute = await this.userRouteRepository.create(createdData);
+
+		return createdRoute.toObject();
+	}
+
+	public async finish(payload: {
+		actualGeometry: LineStringGeometry;
+		routeId: number;
+		userId: number;
+	}): Promise<UserRouteResponseDto> {
+		const { actualGeometry, routeId, userId } = payload;
+
+		const userRoute = await this.getByRouteId(routeId);
+
+		this.ensureUserIsOwner(userRoute.userId, userId);
+
+		this.ensureReadyToFinish(userRoute.status);
+
+		const updatedData = UserRouteEntity.initialize({
+			...userRoute,
+			actualGeometry,
+			completedAt: getCurrentIsoDate(),
+			status: UserRouteStatus.COMPLETED,
+		});
+
+		const updatedRoute = (await this.userRouteRepository.patch(
+			userRoute.id,
+			userId,
+			updatedData,
+		)) as UserRouteEntity;
```

Comment:
It's better not to cast here, because during patching, we may pass filters that may have been overlooked, which will eventually return null. So it's better to handle this case here and throw a not found error
```suggestion
		));
```